indexing-service/src/test/java/org/apache/druid/indexing/common/task/CompactionTaskParallelRunTest.java

```diff
@@ -0,0 +1,320 @@
+import org.apache.druid.timeline.DataSegment;
+import org.apache.druid.timeline.SegmentId;
+import org.joda.time.Interval;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import javax.annotation.Nullable;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+
+@RunWith(Parameterized.class)
+public class CompactionTaskParallelRunTest extends AbstractParallelIndexSupervisorTaskTest
+{
+  @Parameterized.Parameters(name = "{0}")
+  public static Iterable<Object[]> constructorFeeder()
+  {
+    return ImmutableList.of(
+        new Object[]{LockGranularity.TIME_CHUNK},
+        new Object[]{LockGranularity.SEGMENT}
+    );
+  }
+
+  private static final String DATA_SOURCE = "test";
+  private static final RetryPolicyFactory RETRY_POLICY_FACTORY = new RetryPolicyFactory(new RetryPolicyConfig());
+
+  private final AppenderatorsManager appenderatorsManager = new TestAppenderatorsManager();
+  private final LockGranularity lockGranularity;
+  private final RowIngestionMetersFactory rowIngestionMetersFactory;
+  private final CoordinatorClient coordinatorClient;
+
+  public CompactionTaskParallelRunTest(LockGranularity lockGranularity)
+  {
+    this.lockGranularity = lockGranularity;
+    this.rowIngestionMetersFactory = new TestUtils().getRowIngestionMetersFactory();
+    coordinatorClient = new CoordinatorClient(null, null)
+    {
+      @Override
+      public List<DataSegment> getDatabaseSegmentDataSourceSegments(String dataSource, List<Interval> intervals)
+      {
+        return getStorageCoordinator().getUsedSegmentsForIntervals(dataSource, intervals);
+      }
+
+      @Override
+      public DataSegment getDatabaseSegmentDataSourceSegment(String dataSource, String segmentId)
+      {
+        ImmutableDruidDataSource druidDataSource = getMetadataSegmentManager().getImmutableDataSourceWithUsedSegments(
+            dataSource
+        );
+        if (druidDataSource == null) {
+          throw new ISE("Unknown datasource[%s]", dataSource);
+        }
+
+        for (SegmentId possibleSegmentId : SegmentId.iteratePossibleParsingsWithDataSource(dataSource, segmentId)) {
+          DataSegment segment = druidDataSource.getSegment(possibleSegmentId);
+          if (segment != null) {
+            return segment;
+          }
+        }
+        throw new ISE("Can't find segment for id[%s]", segmentId);
+      }
+    };
+  }
+
+  @Before
+  public void setup() throws IOException
+  {
+    indexingServiceClient = new LocalIndexingServiceClient();
+    localDeepStorage = temporaryFolder.newFolder();
+  }
+
+  @After
+  public void teardown()
+  {
+    indexingServiceClient.shutdown();
+    temporaryFolder.delete();
+  }
+
+  @Test
+  public void test() throws Exception
```

Comment:
Thanks, missed renaming it. Changed to `testRunParallel` now.

> Do any of the test scenarios in `CompactionTaskRunTest` make sense here? If so, maybe some code can be factored. For example, the `runIndexTask()` method in both classes is similar.

Good point. I think your comment has a point, but the best way I'm think of is adding a new framework which can run any kind of tasks in unit tests which is similar to ingestion test framework but debuggable. With this framework, we don't have to write a lot of boilerplate codes to set up an environment to run tasks. `AbstractParallelIndexSupervisorTaskTest` is a sort of that framework, but it's only for parallel indexing tasks and not really easy to write tests.

Regarding `runIndexTask()`, I guess we can extract that method as a util method, but I'm not sure if it's worth because `runIndexTask()` is to prepare tests by adding some data before compaction and test data can be changed at any time.